test/csuite/wt4156_metadata_salvage/main.c

```diff
@@ -0,0 +1,462 @@
+	end = buf + bufsize;
+	s = NULL;
+	c = (int)str[0];
+	buflen = bufsize;
+	slen = strlen(str);
+	/*
+	 * Find the first character and then compare.
+	 */
+	while ((s = memchr(p, c, buflen)) != NULL) {
+		/*
+		 * If we don't have enough buffer left to compare we do not
+		 * have a match.
+		 */
+		buflen = (size_t)(end - s);
+		if (buflen < slen)
+			return (NULL);
+		if (memcmp(s, str, slen) == 0)
+			return (s);
+		/*
+		 * This one didn't match, increment in the buffer and find the
+		 * next one.
+		 */
+		++s;
+		p = s;
+	}
+	return (NULL);
+}
+
+/*
+ * cursor_insert --
+ *	Insert some data into a table.
+ */
+static void
+cursor_insert(WT_SESSION *session, const char *uri)
+{
+	WT_CURSOR *cursor;
+	WT_ITEM vu;
+	char keybuf[100], valuebuf[100];
+	bool recno;
+
+	/* Reserve requires a running transaction. */
+	testutil_check(session->begin_transaction(session, NULL));
+
+	memset(&vu, 0, sizeof(vu));
+
+	/* Open a cursor. */
+	testutil_check(session->open_cursor(session, uri, NULL, NULL, &cursor));
+
+	/* Operations change based on the key/value formats. */
+	recno = strcmp(cursor->key_format, "r") == 0;
+	if (recno)
+		cursor->set_key(cursor, (uint64_t)1);
+	else {
+		strcpy(keybuf, KEY);
+		cursor->set_key(cursor, keybuf);
+	}
+	strcpy(valuebuf, VALUE);
+	cursor->set_value(cursor, valuebuf);
+	testutil_check(cursor->insert(cursor));
+	testutil_check(cursor->close(cursor));
+}
+
+/*
+ * create_data --
+ * 	Create a table and insert a piece of data.
+ */
+static void
+create_data(WT_CONNECTION *conn, TABLE_INFO *t)
+{
+	WT_SESSION *session;
+	size_t len;
+	uint64_t i;
+	char buf[APP_BUF_SIZE], cfg[APP_MD_SIZE];
+
+	memset(buf, 0, sizeof(buf));
+	memset(cfg, 0, sizeof(cfg));
+
+	/*
+	 * Create an app-specific metadata string that fills most of page
+	 * so that each table in the metadata has its own page.
+	 */
+	len = strlen(APP_STR);
+	for (i = 0; i + len < APP_BUF_SIZE; i += len)
+		testutil_check(__wt_snprintf(
+		    &buf[i], APP_BUF_SIZE - i, "%s", APP_STR));
+	testutil_check(__wt_snprintf(cfg, sizeof(cfg),
+	    "%s,app_metadata=\"%s\"", t->kvformat, buf));
+	testutil_check(conn->open_session(conn, NULL, NULL, &session));
+	testutil_check(session->create(session, t->name, cfg));
+	cursor_insert(session, t->name);
+	testutil_check(session->close(session, NULL));
+}
+
+/*
+ * corrupt_metadata --
+ *	Corrupt the metadata by scribbling on the "corrupt" URI string.
+ */
+static void
+corrupt_metadata(const char *home)
+{
+	FILE *fp;
+	struct stat sb;
+	long off;
+	size_t meta_size;
+	bool corrupted;
+	uint8_t *buf, *corrupt;
+	char path[256];
+
+	/*
+	 * Open the file, read its contents. Find the string "corrupt" and
+	 * modify one byte at that offset. That will cause a checksum error
+	 * when WiredTiger next reads it.
+	 */
+	sprintf(path, "%s/%s", home, WT_METAFILE);
+	testutil_check(stat(path, &sb));
+	meta_size = (size_t)sb.st_size;
+	buf = dmalloc(meta_size);
+	memset(buf, 0, meta_size);
+	if ((fp = fopen(path, "r+")) == NULL)
```

Comment:
I think we should have the `'b'` option for `fopen`.  Not sure if this will ever run on Windows, but just in case.